test/Npgsql.Tests/Types/JsonDynamicTests.cs

```diff
@@ -215,110 +215,124 @@ await AssertType(
-    }
+        var value = new ExtendedDerivedWeatherForecast
+        {
+            Date = new DateTime(2019, 9, 1),
+            Summary = "Partly cloudy",
+            TemperatureC = 10
+        };
+
+        // Note: we assert a specific string representation, though jsonb doesn't guarantee the property ordering; so the assert may break
+        // for jsonb if PostgreSQL changes its implementation.
+        var sql =
+            IsJsonb
+                ? """{"Date": "2019-09-01T00:00:00", "$type": "extended", "Summary": "Partly cloudy", "TemperatureC": 10, "TemperatureF": 49}"""
+                : """{"$type":"extended","TemperatureF":49,"Date":"2019-09-01T00:00:00","TemperatureC":10,"Summary":"Partly cloudy"}""";
 
+        await AssertTypeWrite(dataSource, value, sql, PostgresType, NpgsqlDbType, isNpgsqlDbTypeInferredFromClrType: false);
+        await AssertTypeRead(dataSource, sql, PostgresType, value, isDefault: false);
+    }
 
     [Test]
-    public async Task Poco_exact_polymorphic_mapping()
+    public async Task Poco_unspecified_polymorphic_mapping()
     {
-        // We don't yet support polymorphic deserialization for jsonb as $type does not come back as the first property.
-        // This could be fixed by detecting PolymorphicOptions types, always buffering their values and modifying the text.
-        if (IsJsonb)
-            return;
+        await using var dataSource = CreateWeatherForecastDataSource();
 
-        var dataSourceBuilder = CreateDataSourceBuilder();
-        dataSourceBuilder.EnableDynamicJson(jsonClrTypes: [typeof(ExtendedDerivedWeatherForecast)]);
-        await using var dataSource = dataSourceBuilder.Build();
+        var value = new ExtendedDerivedWeatherForecast
+        {
+            Date = new DateTime(2019, 9, 1),
+            Summary = "Partly cloudy",
+            TemperatureC = 10
+        };
 
-        await AssertType(
-            dataSource,
-            new ExtendedDerivedWeatherForecast()
+        // Note: we assert a specific string representation, though jsonb doesn't guarantee the property ordering; so the assert may break
+        // for jsonb if PostgreSQL changes its implementation.
+        var sql =
+            IsJsonb
+                ? """{"Date": "2019-09-01T00:00:00", "$type": "extended", "Summary": "Partly cloudy", "TemperatureC": 10, "TemperatureF": 49}"""
+                : """{"$type":"extended","TemperatureF":49,"Date":"2019-09-01T00:00:00","TemperatureC":10,"Summary":"Partly cloudy"}""";
+
+        await AssertTypeWrite(dataSource, value, sql, PostgresType, NpgsqlDbType, isDefault: false);
+
+        await AssertTypeRead(dataSource, sql, PostgresType,
+            new DerivedWeatherForecast
             {
                 Date = new DateTime(2019, 9, 1),
                 Summary = "Partly cloudy",
                 TemperatureC = 10
             },
-            """{"TemperatureF":49,"Date":"2019-09-01T00:00:00","TemperatureC":10,"Summary":"Partly cloudy"}""",
```

Comment:
I'm not too familiar with STJ polymorphism, but Is it OK that we didn't have `$type` here previously (looks like a possible bug)? This is json (not jsonb), and we're writing a type in a hierarchy without a discriminator?

Reply:
I think it's because the derived type is being serialized directly instead of serializing it as the base type; i.e. doing `AssertTypeWrite<WeatherForecast>` instead of `AssertTypeWrite<ExtendedDerivedWeatherForecast>`

Reply:
@roji yep it's a bit confusing but the rationale given by the STJ devs is that the exact type shouldn't need a $type discriminator. 

Effectively the behavior is to deserialize a string to the type that was requested *unless* a $type discriminator is found that matches a derived type  (specified via the attribute, or via the contract etc.) on the type you're deserializing. If those things line up STJ will create an instance of the derived type instead.

For serialization this is mostly the inverse (serializing a derived instance as the defining base type causes $type to be included), except for when typeof(object) as the type is specified during serialization. For typeof(object) STJ uses the runtime type to serialize, but it differs from passing just the value (without a specific base type) in that it also supports polymorphism by finding the nearest ancestor that specifies the type (throwing on ambiguity etc). 

Simple right? :D

Reply:
Yeah, from a pure STJ standpoint I can kinad see how that makes sense, but I'm wondering what's right in Npgsql specifically. Like if someone sends a parameter to the database that has its value set to a derived class, it (strongly) feels like `$type` should be written there, otherwise important information is lost... The crucial thing is that in most Npgsql usage, NpgsqlParameter is non-generic (that's just how ADO.NET works) and therefore there's no base type for the serialization... In other words I don't think we should require people to use the Npgsql-specific `NpgsqlParameter<T>` in order to get a `$type` property in the database (which IIUC is the case before this PR at least?).

Reply:
Not having a $type on the serialized value seems like a limitation upstream which exists regardless of this PR -- unless you propose to fix it here somehow? Not sure how without source generators or reflection (like crawling the base types for attributes?)

Reply:
> Yeah, from a pure STJ standpoint I can kinad see how that makes sense, but I'm wondering what's right in Npgsql specifically. Like if someone sends a parameter to the database that has its value set to a derived class, it (strongly) feels like `$type` should be written there, otherwise important information is lost... The crucial thing is that in most Npgsql usage, NpgsqlParameter is non-generic (that's just how ADO.NET works) and therefore there's no base type for the serialization... In other words I don't think we should require people to use the Npgsql-specific `NpgsqlParameter<T>` in order to get a `$type` property in the database (which IIUC is the case before this PR at least?).

That's not what's happening no, things work as expected (within the STJ model) in 8.0 for db type json. We broadly have two cases to consider in JsonDynamicResolver: registered and unregistered (de)serialization.

If you don't register the type (i.e. you don't pass it to `EnableDynamicJson`) we use `typeof(object)` as the base type for json (and the exact type for jsonb, the condition we remove in this PR), ASP.NET's json output formatter does the same (and they were actually the reason `typeof(object)` handling was added to STJ for 8.0). As mentioned, this will try to find the nearest ancestor (or itself) with polymorphic type info and output $type accordingly. NpgsqlParameter then works as expected in almost all cases, more on that below. 

If you do register the type there are a few extra things that happen:
- We add the type as a json/jsonb mapping which means you don't need to specify a datatypename (which was the original intent behind the registrations, before the 8.0 rewrite)
- From 8.0 onwards we additionally check for each registration type whether there are any `PolymorphicOptions.DerivedTypes`, if so we add all these types as mappings as well. This is done precisely to support polymorphism and runtime typed dbparameters. For each of these derived types we make sure we return a jsonconverter that uses the registration type as `TBase`, so that `$type` is emitted.

The latter mechanism also serves as an escape hatch for the ambiguity cases you can run into with typeof(object). A simple example would be when the instance type implements both a base type and an interface that define derivedtype information. In such cases STJ will throw and you could be stuck. if you instead register which of the two base types you want in Npgsql we'll make sure to set the right base type and make it work. 

It's also possible to use `NpgsqlParameter<T>` to set the base type you want (while setting some derived value), but as you already noted generic NpgsqlParameter is npgsql specific and library support is poor.

Reply:
OK, as @NinoFloris pointed out to me offline, this was my bad: the discriminator value needs to be specified explicitly:

```c#
[JsonDerivedType(typeof(Child), typeDiscriminator: "child")]
```

Once that's done, `$type` gets emitted.